daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package speedy
+
+import com.daml.lf.transaction.{NodeId, GenTransaction}
+import com.daml.lf.transaction.Node.{
+  GenNode,
+  GenActionNode,
+  NodeRollback,
+  NodeExercises,
+  LeafOnlyActionNode,
+}
+import com.daml.lf.value.Value
+import com.daml.lf.data.ImmArray
+
+private[lf] object NormalizeRollbacks {
+
+  type Nid = NodeId
+  type Cid = Value.ContractId
+  type TX = GenTransaction[Nid, Cid]
+  type Node = GenNode[Nid, Cid]
+  type ActNode = GenActionNode[Nid, Cid]
+
+  // Normalize a transaction so rollback nodes satisfy the normalization rules.
+  // see `makeRoll` below
+
+  def normalizeTx(txOriginal: TX): TX = {
+
+    // Here we generate fresh node-ids for the normalized transaction.
+    val ids = Iterator.from(0).map(NodeId(_)) // perhaps we could reuse orig nids
+
+    // There is no connection between the ids in the pre and post transaction.
+    // Semantically this is fine: The ids are just of tie the tree structure together.
+    // But perhaps for ease of debugging we might consider trying to reuse existing ids.
+
+    // We accumulate (imperatively!) a nodes-map for the normalized transaction
+    var nodes: Map[Nid, Node] = Map.empty
+
+    def pushNode(node: Node): Nid = {
+      val nodeId = ids.next()
+      nodes += (nodeId -> node)
+      nodeId
+    }
+
+    def pushNodes(nodes: List[Node]): List[Nid] = nodes.map(pushNode)
+
+    import Canonical.{Norm, Case, caseNorms}
+    // The normalization phase works by constructing intermediate values which are
+    // `Canonical` in the sense that only properly normalized nodes can be represented.
+    // Although this doesn't ensure correctness, one class of bugs is avoided.
+
+    // The `force*` functions move from the world of Norms into standard transactions.
+    // This is also where node-ids for the resulting normalized transaction are generated.
+
+    def forceRoll(x: Norm.Roll): Node = {
+      x match {
+        case Norm.Roll1(act) =>
+          val child = pushNode(act.node)
```

Comment:
the logic around when we call `forceRoll` definitely looks like it doesn’t result in `pushNode` calls in prefix order either.

Reply:
fixed